Correct order of overflowing tokens for LayoutLmV2 tokenizer

Apoorvgarg-creator opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->
This PR objective is same as #13179 
Fixes #13148
The issue was resolved for every tokenizer except the LayoutLmV2 tokenizer.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section? Yes 👍🏻.
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Yes👍🏻 , #13148
- [x] Did you make sure to update the documentation with your changes? Yes 👍🏻.
- [x] Did you add any necessary tests?
The following test have been added which check the sequence of **`overflowing tokens`**, **`bbox sequence`** and **`input_ids`**.
     - test_maximum_encoding_length_pair_input
     - test_maximum_encoding_length_single_input 
     

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->
@LysandreJik @sgugger @SaulLu @NielsRogge

Apoorvgarg-creator committed:
correct order of overflowing tokens for LayoutLmV2 tokenizer

SaulLu commented:
Thank you very much for the PR. 

As it seems to me that we didn't see any tests failing for LayoutLMv2 at the time your previous PR was merged, could you confirm that this behaviour is tested for LayoutLMv2 now? As I see that in the PR you did not change any tests, I'd like to make sure we catch the problem with the tests next time :slightly_smiling_face: .

Apoorvgarg-creator commented:
> As I see that in the PR you did not change any tests, I'd like to make sure we catch the problem with the tests next time 🙂 .

Sure.
Thank you

Apoorvgarg-creator commented:
@SaulLu, I have been trying to resolve the error, 
I have made a function `get_clean_sequences` as follow-
### Changes -
```
    def get_clean_sequence(self, tokenizer, with_prefix_space=False, max_length=20, min_length=5):
        toks = [(i, tokenizer.decode([i], clean_up_tokenization_spaces=False)) for i in range(len(tokenizer))]
        toks = list(filter(lambda t: re.match(r"^[ a-zA-Z]+$", t[1]), toks))
        toks = list(filter(lambda t: [t[0]] == tokenizer.encode(t[1].split(" "), boxes=len(t[1]) * [[1, 1, 1, 1]], add_special_tokens=False), toks))
        if max_length is not None and len(toks) > max_length:
            toks = toks[:max_length]
        if min_length is not None and len(toks) < min_length and len(toks) > 0:
            while len(toks) < min_length:
                toks = toks + toks
        # toks_str = [t[1] for t in toks]
        toks_ids = [t[0] for t in toks]

        # Ensure consistency
        output_txt = tokenizer.decode(toks_ids, clean_up_tokenization_spaces=False)
        if " " not in output_txt and len(toks_ids) > 1:
            output_txt = (
                tokenizer.decode([toks_ids[0]], clean_up_tokenization_spaces=False)
                + " "
                + tokenizer.decode(toks_ids[1:], clean_up_tokenization_spaces=False)
            )
        if with_prefix_space:
            output_txt = " " + output_txt
        output_txt_list = output_txt.split(" ")
        output_ids = tokenizer.encode(output_txt_list, boxes=len(output_txt) * [[1, 1, 1, 1]], add_special_tokens=False)
        boxes = len(output_txt) * [[1, 1, 1, 1]]
        return output_txt_list, output_ids, boxes
```

```
 def test_maximum_encoding_length_single_input(self):
        tokenizers = self.get_tokenizers(do_lower_case=False, model_max_length=100)
        for tokenizer in tokenizers:
            with self.subTest(f"{tokenizer.__class__.__name__}"):
                seq_0, ids, boxes = self.get_clean_sequence(tokenizer, max_length=20)

                sequence = tokenizer.encode(seq_0,boxes=boxes, add_special_tokens=False)
                total_length = len(sequence)

                assert total_length > 4, "Issue with the testing sequence, please update it it's too short"

                # Test with max model input length
                model_max_length = tokenizer.model_max_length
                self.assertEqual(model_max_length, 100)
                seq_1 = seq_0 * model_max_length

                sequence1 = tokenizer(seq_1, boxes=len(seq_1)*[[1,1,1,1]], add_special_tokens=False)
                total_length1 = len(sequence1["input_ids"])
                assert (
                    total_length1 > model_max_length
                ), "Issue with the testing sequence, please update it it's too short"

                # Simple
                padding_strategies = (
                    [False, True, "longest"] if tokenizer.pad_token and tokenizer.pad_token_id >= 0 else [False]
                )
                for padding_state in padding_strategies:
                    with self.subTest(f"Padding: {padding_state}"):
                        for truncation_state in [True, "longest_first", "only_first"]:
                            with self.subTest(f"Truncation: {truncation_state}"):
                                output = tokenizer(seq_1,boxes=len(seq_1)*[[1,1,1,1]], padding=padding_state, truncation=truncation_state)
                                self.assertEqual(len(output["input_ids"]), model_max_length)

                                output = tokenizer([seq_1],boxes=[len(seq_1)*[[1,1,1,1]]], padding=padding_state, truncation=truncation_state)
                                self.assertEqual(len(output["input_ids"][0]), model_max_length)

                        # Simple with no truncation
                        # Reset warnings
                        tokenizer.deprecation_warnings = {}
                        with self.assertLogs("transformers", level="WARNING") as cm:
                            output = tokenizer(seq_1, boxes=len(seq_1)*[[1,1,1,1]], padding=padding_state, truncation=False)
                            self.assertNotEqual(len(output["input_ids"]), model_max_length)
                        self.assertEqual(len(cm.records), 1)
                        self.assertTrue(
                            cm.records[0].message.startswith(
                                "Token indices sequence length is longer than the specified maximum sequence length for this model"
                            )
                        )

                        tokenizer.deprecation_warnings = {}
                        with self.assertLogs("transformers", level="WARNING") as cm:
                            output = tokenizer([seq_1],boxes=[len(seq_1)*[[1,1,1,1]]], padding=padding_state, truncation=False)
                            self.assertNotEqual(len(output["input_ids"][0]), model_max_length)
                        self.assertEqual(len(cm.records), 1)
                        self.assertTrue(
                            cm.records[0].message.startswith(
                                "Token indices sequence length is longer than the specified maximum sequence length for this model"
                            )
                        )

                # Overflowing tokens
                stride = 2
                information = tokenizer(
                    seq_0,
                    boxes = len(seq_0) * [[1, 1, 1, 1]],
                    max_length=total_length - 2,
                    add_special_tokens=False,
                    stride=stride,
                    truncation=True,
                    return_overflowing_tokens=True,
                    # add_prefix_space=False,
                )

                truncated_sequence = information["input_ids"]
                overflowing_tokens = information["overflowing_tokens"]
                self.assertEqual(len(truncated_sequence), total_length - 2)
                self.assertEqual(truncated_sequence, sequence[:-2])
                self.assertEqual(len(overflowing_tokens), 2 + stride)
                self.assertEqual(overflowing_tokens, sequence[-(2 + stride) :])
```


I ran the function and got the error that **In the dictionary `information` overflowing_tokens key doesnot exist**
### Error Message/Console Output -
```
self = {'input_ids': [[5, 6, 7, 8, 10, 11], [10, 11, 13, 14]], 'token_type_ids': [[0, 0, 0, 0, 0, 0], [0, 0, 0, 0]], 'attenti...1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]], [[1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]]]}
item = 'overflowing_tokens'

    def __getitem__(self, item: Union[int, str]) -> Union[Any, EncodingFast]:
        """
        If the key is a string, returns the value of the dict associated to :obj:`key` ('input_ids', 'attention_mask',
        etc.).
    
        If the key is an integer, get the :obj:`tokenizers.Encoding` for batch item with index :obj:`key`.
        """
        if isinstance(item, str):
>           return self.data[item]
E           KeyError: 'overflowing_tokens'

src/transformers/tokenization_utils_base.py:237: KeyError
```

Apoorvgarg-creator committed:
test to check order of overflowing_tokens for a seq of input_ids

Apoorvgarg-creator committed:
fix up quality

Apoorvgarg-creator commented:
@SaulLu, I have added a function `get_clean_sequences` which gives us list of words along with their ids and boxes, have taken reference from the `test_tokenization_common.py`.
Furthermore, I have modified the `test_maximum_encoding_length_single_input` to ensure the correct order of overflowing order.
Please can you review the changes.
Thank you

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+        tokenizers = self.get_tokenizers(do_lower_case=False, model_max_length=100)
+        for tokenizer in tokenizers:
+            with self.subTest(f"{tokenizer.__class__.__name__}"):
+                seq_0, ids, boxes = self.get_clean_sequence(tokenizer, max_length=20)
```
I propose here to also number the list for the boxes to lighten the rest of the code (see next comment)

```suggestion
                seq_0, boxes_0, ids = self.get_clean_sequence(tokenizer, max_length=20)
```

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+                # Test with max model input length
+                model_max_length = tokenizer.model_max_length
+                self.assertEqual(model_max_length, 100)
+                seq_1 = seq_0 * model_max_length
```
To lighten the code, I propose to make the following change, which will then replace all occurrences of `len(seq_1) * [[1, 1, 1, 1]]` by `boxes_1`

```suggestion
                seq_1 = seq_0 * model_max_length
                boxes_1 = boxes_0 * model_max_length
```

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+                                    padding=padding_state,
+                                    truncation=truncation_state,
+                                )
+                                self.assertEqual(len(output["input_ids"]), model_max_length)
```
As LayoutLmV2 differs from other tokenizers by introducing boxes, I think it would be nice to also add tests to check the length of these sequences too.
```suggestion
                                self.assertEqual(len(output["input_ids"]), model_max_length)
                                self.assertEqual(len(output["bbox"]), model_max_length)
```

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+                                    padding=padding_state,
+                                    truncation=truncation_state,
+                                )
+                                self.assertEqual(len(output["input_ids"][0]), model_max_length)
```
idem
```suggestion
                                self.assertEqual(len(output["input_ids"][0]), model_max_length)
                                self.assertEqual(len(output["bbox"][0]), model_max_length)
```

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+                            output = tokenizer(
+                                seq_1, boxes=len(seq_1) * [[1, 1, 1, 1]], padding=padding_state, truncation=False
+                            )
+                            self.assertNotEqual(len(output["input_ids"]), model_max_length)
```
again :) 
```suggestion
                            self.assertNotEqual(len(output["input_ids"]), model_max_length)
                            self.assertNotEqual(len(output["bbox"]), model_max_length)
```

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+                            output = tokenizer(
+                                [seq_1], boxes=[len(seq_1) * [[1, 1, 1, 1]]], padding=padding_state, truncation=False
+                            )
+                            self.assertNotEqual(len(output["input_ids"][0]), model_max_length)
```
same
```suggestion
                            self.assertNotEqual(len(output["input_ids"][0]), model_max_length)
                            self.assertNotEqual(len(output["bbox"][0]), model_max_length)
```

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1781,9 +1789,150 @@ def test_alignement_methods(self):
+                )
+
+                # Overflowing tokens are handled quite differently in slow and fast tokenizers
+                if isinstance(tokenizer, PreTrainedTokenizerFast):
+                    truncated_sequence = information["input_ids"][0]
+                    overflowing_tokens = information["input_ids"][1]
+                    self.assertEqual(len(information["input_ids"]), 2)
+
+                    self.assertEqual(len(truncated_sequence), total_length - 2)
+                    self.assertEqual(truncated_sequence, sequence[:-2])
+
+                    self.assertEqual(len(overflowing_tokens), 2 + stride)
+                    self.assertEqual(overflowing_tokens, sequence[-(2 + stride) :])
+                else:
+                    truncated_sequence = information["input_ids"]
+                    overflowing_tokens = information["overflowing_tokens"]
+
+                    self.assertEqual(len(truncated_sequence), total_length - 2)
+                    self.assertEqual(truncated_sequence, sequence[:-2])
+
+                    self.assertEqual(len(overflowing_tokens), 2 + stride)
+                    self.assertEqual(overflowing_tokens, sequence[-(2 + stride) :])
```
It would be really nice to test also that everything goes well for the `bbox` sequence (which is a particularity of layoutlmv2). This would mean that the line `sequence = tokenizer.encode(seq_0, boxes=boxes, add_special_tokens=False)` isn't sufficient to retrieve the ground truth here, you will probably need to use  `tokenizer(seq_0, boxes=boxes, add_special_tokens=False)`.

SaulLu reviewed (commented):
Thank you very much for all this work @Apoorvgarg-creator. :confetti_ball: 

I left you mostly comments about the testing part. Especially as LayoutLmV2 has extra boxes, I think we should also add checks on the corresponding sequences. 

Finally, if I'm not mistaken, there is no `test_maximum_encoding_length_pair_input` test for lajoutlmv2, it would be nice to have it too. :smile: 

Let me know if you need some help!

Apoorvgarg-creator commented:
@SaulLu, I have gone through the reviews. I will make the changes.
I didn't commit `test_maximum_encoding_length_pair_input` function because I wanted to see the changes that are required in the `get_clean_sequences` and `test_maximum_encoding_length_single_input`.
Thank you for reviewing and suggesting the changes.

Apoorvgarg-creator commented:
> It would be really nice to test also that everything goes well for the bbox sequence (which is a particularity of layoutlmv2). This would mean that the line sequence = tokenizer.encode(seq_0, boxes=boxes, add_special_tokens=False) isn't sufficient to retrieve the ground truth here, you will probably need to use tokenizer(seq_0, boxes=boxes, add_special_tokens=False).

@SaulLu, I have resolved all the previous suggestions.
Can you explain a bit more about this?

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

Apoorvgarg-creator committed:
added suggested changes

Apoorvgarg-creator committed:
Merge branch 'master' into LayoutLmV2-overflowing-tokens-order-fix

Apoorvgarg-creator committed:
check that tests the bbox sequence

Apoorvgarg-creator committed:
pair_input test added

Apoorvgarg-creator committed:
pass quality test

Apoorvgarg-creator committed:
check bbox sequence added

Apoorvgarg-creator commented:
@SaulLu @LysandreJik @NielsRogge, could you please review the test added and changes.
Thank you

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

Apoorvgarg-creator commented:
@patrickvonplaten @SaulLu @LysandreJik @NielsRogge , Could you please review the PR.
Thank you.

LysandreJik commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1777,13 +1778,481 @@ def test_batch_encode_dynamic_overflowing(self):
+                    ids = None
+
+                seq0_tokens = tokenizer(seq_0, boxes=boxes_0, add_special_tokens=False)
+                assert len(seq0_tokens["input_ids"]) > 2 + stride
```
Could you replace the `assert` with `unittest`'s `self.assertXXX`?

It gives better information on the tests failing.

Apoorvgarg-creator replied:
Okay, I will make the changes.
Thank you

LysandreJik reviewed (commented):
Thank you for the very extensive tests, great work! 
Regarding the format:
- Could you add some comments explaining what is done? There is a lot of code so I fear it will become unmaintainable without clear comments explaining what is done.
- Please replace all `assert` with unittest's specific methods.

I'll let @NielsRogge review the logic.

Apoorvgarg-creator commented:
> * Could you add some comments explaining what is done? There is a lot of code so I fear it will become unmaintainable without clear comments explaining what is done.
Sure, Thank you. for reviewing.

NielsRogge commented on src/transformers/models/layoutlmv2/tokenization_layoutlmv2.py:
```diff
@@ -893,7 +893,9 @@ def prepare_for_model(
         Prepares a sequence or a pair of sequences so that it can be used by the model. It adds special tokens,
         truncates sequences if overflowing while taking into account the special tokens and manages a moving window
-        (with user defined stride) for overflowing tokens.
+        (with user defined stride) for overflowing tokens. Please Note, for `text_pair` different than `None` and
+        `truncation_strategy = longest_first` or `True`, it is not possible to return overflowing tokens. Such a
+        combination of arguments will raise an error.
```
What's the reason the `longest_first` strategy cannot be used when providing a text pair?

Apoorvgarg-creator replied:
This was discussed in the PR #13179 
Three Strategies were proposed in the above PR, [Comment](https://github.com/huggingface/transformers/pull/13179#issuecomment-902669114)

Conclusion Drawn :
https://github.com/huggingface/transformers/pull/13179#issuecomment-906566964


So, I Thought same case will happen with the LayoutLmv2 as well.

Apoorvgarg-creator committed:
unittest method

Apoorvgarg-creator committed:
Merge branch 'LayoutLmV2-overflowing-tokens-order-fix' of https://github.com/Apoorvgarg-creator/transformers into LayoutLmV2-overflowing-tokens-order-fix

NielsRogge reviewed (approved):
This PR looks good to me.

Pinging @SaulLu for a final review.

Apoorvgarg-creator committed:
comments added

SaulLu committed:
add overflowing bbox test

SaulLu commented on tests/test_tokenization_layoutlmv2.py:
```diff
@@ -1777,13 +1778,482 @@ def test_batch_encode_dynamic_overflowing(self):
+                seq0_tokens = tokenizer(seq_0, boxes=boxes_0, add_special_tokens=False)
+                self.assertGreater(len(seq0_tokens["input_ids"]), 2 + stride)
+                question_1 = "This is another sentence to be encoded."
+                seq_1 = ["This", "is", "another", "sentence", "to", "be", "encoded."]
```
If I'm not mistaking `seq1_tokens['input_ids']` (defined 2 lines below) is equal to `[0, 0, 0, 0, 0, 0, 0, 0]` which correspond to a list of unknown tokens. It is due to the fact that we use a toy tokenizer for the tests. Maybe we can change it a bit in order to have something which is not only composed of unknown tokens? and even an example where a word is divided into 2 tokens (in the following example it is the word "weirdly")
```suggestion
                seq_1 = ["what", "a", "weird", "test", "weirdly", "weird"]
```

SaulLu reviewed (commented):
Thank you very much for all your work @Apoorvgarg-creator !

Thanks especially for adding all those tests! I have one last little request which is about the overflowing boxes check, but to save you time I have [opened a PR](https://github.com/Apoorvgarg-creator/transformers/pull/2) on your branch which includes the changes I am referring to. If you like these proposals, don't hesitate to merge this PR in your branch. :slightly_smiling_face:

Apoorvgarg-creator committed:
improved "seq_1"

Co-authored-by: SaulLu <55560583+SaulLu@users.noreply.github.com>

Apoorvgarg-creator committed:
Merge pull request #2 from SaulLu/add-overflowing-bbox-test

add overflowing bbox test

Apoorvgarg-creator committed:
Merge branch 'huggingface:master' into LayoutLmV2-overflowing-tokens-order-fix

pggee7198 commented on src/transformers/models/layoutlmv2/tokenization_layoutlmv2.py:
```diff
@@ -650,7 +650,7 @@ def _batch_prepare_for_model(
         """
         Prepares a sequence of input id, or a pair of sequences of inputs ids so that it can be used by the model. It
         adds special tokens, truncates sequences if overflowing while taking into account the special tokens and
-        manages a moving window (with user defined stride) for overflowing tokens
```
gzla

Apoorvgarg-creator committed:
improve code quality

SaulLu reviewed (approved):
Thank you for these latest changes! It's all good for me!

Apoorvgarg-creator commented:
@SaulLu @LysandreJik, I have been getting these mails, I tried checking the reason for their failing but couldn't understand the reason for failing.
Here the screenshot -
![IMG_6135](https://user-images.githubusercontent.com/57873504/140864209-d8486bf5-1151-48e0-8e7f-af9692759d30.PNG)

sgugger commented:
Thnaks a gain for all your work on this!
